File: Samples/java/src/main/java/com/phunware/java/sample/NavigationOverlayView.java
package com.phunware.java.sample;

/* Copyright (C) 2018 Phunware, Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL Phunware, Inc. BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of Phunware, Inc. shall
not be used in advertising or otherwise to promote the sale, use or
other dealings in this Software without prior written authorization
from Phunware, Inc. */

import android.content.Context;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.phunware.core.PwLog;
import com.phunware.mapping.manager.Navigator;
import com.phunware.mapping.model.BuildingOptions;
import com.phunware.mapping.model.PointOptions;
import com.phunware.mapping.model.RouteManeuverOptions;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class NavigationOverlayView extends ViewPager
        implements Navigator.OnManeuverChangedListener {

    private Navigator navigator;
    private ManeuverPagerAdapter adapter;

    private final OnPageChangeListener pageChangeListener = new SimpleOnPageChangeListener() {
        @Override
        public void onPageSelected(int position) {
            if (navigator != null) {
                ManeuverPair pair = adapter.getItem(position);
                navigator.setCurrentManeuver(pair.mainPos);
            }
        }
    };

    public NavigationOverlayView(Context context) {
        this(context, null);
    }

    public NavigationOverlayView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public void setNavigator(Navigator navigator) {
        this.navigator = navigator;
        navigator.addOnManeuverChangedListener(this);

        List<ManeuverPair> pairs = new ArrayList<>();
        List<RouteManeuverOptions> maneuvers = this.navigator.getManeuvers();
        for (int i = 0; i < maneuvers.size(); i += 2) {
            ManeuverPair pair = new ManeuverPair();
            pair.mainPos = i;
            pair.mainManeuver = maneuvers.get(i);

            if (i + 1 < maneuvers.size()) {
                RouteManeuverOptions next = maneuvers.get(i + 1);
                if (next.isTurnManeuver() || next.isPortalManeuver()) {
                    pair.turnPos = i + 1;
                    pair.turnManeuver = next;
                } else {
                    i--;
                }
            }

            pairs.add(pair);
        }

        adapter = new ManeuverPagerAdapter(navigator);
        setAdapter(adapter);
        adapter.setManeuvers(pairs);
        addOnPageChangeListener(pageChangeListener);
        setCurrentItem(0);
        navigator.setCurrentManeuver(0);
    }

    @Override
    public void onManeuverChanged(Navigator navigator, int position) {
        for (int i = 0; i < adapter.getCount(); i++) {
            ManeuverPair pair = adapter.getItem(i);
            if (pair.mainPos == position || pair.turnPos == position) {
                setCurrentItem(i);
                return;
            }
        }
    }

    @Override
    public void onRouteSnapFailed() {
        PwLog.e("NavigationOverlayView", "Off route");
    }

    private static class ManeuverPair {
        int mainPos;
        int turnPos;
        RouteManeuverOptions mainManeuver;
        RouteManeuverOptions turnManeuver;
    }

    private static final class ManeuverPagerAdapter extends PagerAdapter {

        private final Navigator navigator;
        private final ManeuverDisplayHelper displayHelper;
        private final List<ManeuverPair> maneuvers = new ArrayList<>();

        ManeuverPagerAdapter(Navigator navigator) {
            super();
            this.navigator = navigator;
            displayHelper = new ManeuverDisplayHelper();
        }

        @Override
        public int getCount() {
            return maneuvers.size();
        }

        void setManeuvers(Collection<? extends ManeuverPair> maneuvers) {
            this.maneuvers.clear();
            this.maneuvers.addAll(maneuvers);
            notifyDataSetChanged();
        }

        ManeuverPair getItem(int position) {
            return maneuvers.get(position);
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            Context context = container.getContext();
            View v = LayoutInflater.from(context)
                    .inflate(R.layout.item_maneuver, container, false);

            ImageView direction = v.findViewById(R.id.direction);
            TextView maneuver = v.findViewById(R.id.maneuver);
            TextView nextManeuver = v.findViewById(R.id.next_maneuver);
            ImageView nextDirection = v.findViewById(R.id.next_direction);

            final ManeuverPair m = maneuvers.get(position);

            direction.setImageResource(displayHelper.getImageResourceForDirection(context,
                    m.mainManeuver));
            maneuver.setText(displayHelper.stringForDirection(context, m.mainManeuver));

            if (m.turnManeuver == null) {
                nextManeuver.setVisibility(View.GONE);
                nextDirection.setVisibility(GONE);
                v.findViewById(R.id.next).setVisibility(GONE);
            } else {
                v.findViewById(R.id.next).setVisibility(VISIBLE);
                nextManeuver.setVisibility(View.VISIBLE);
                nextDirection.setVisibility(VISIBLE);
                nextDirection.setImageResource(displayHelper
                        .getImageResourceForDirection(context, m.turnManeuver));
                nextManeuver.setText(displayHelper.stringForDirection(
                        context, m.turnManeuver));
            }

            if (position == maneuvers.size() - 1) {
                v.findViewById(R.id.next).setVisibility(GONE);
                final int pointCount = navigator.getRoute().getPoints().size();
                PointOptions finalPoint = navigator.getRoute().getPoints().get(pointCount - 1);
                String customLocation = context.getString(R.string.custom_location_title);
                String arrive = context.getString(R.string.to_arrive,
                        finalPoint.getName() == null ? customLocation : finalPoint.getName());
                nextManeuver.setVisibility(View.VISIBLE);
                nextManeuver.setText(arrive);
            }

            container.addView(v);
            return v;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            container.removeView((View) object);
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view.equals(object);
        }
    }
}

File: Samples/kotlin/src/main/java/com/phunware/kotlin/sample/NavigationOverlayView.kt
package com.phunware.kotlin.sample

/* Copyright (C) 2018 Phunware, Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL Phunware, Inc. BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of Phunware, Inc. shall
not be used in advertising or otherwise to promote the sale, use or
other dealings in this Software without prior written authorization
from Phunware, Inc. */

import android.content.Context
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.style.TextAppearanceSpan
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.phunware.core.PwLog
import com.phunware.mapping.manager.Navigator
import com.phunware.mapping.model.RouteManeuverOptions

import java.util.ArrayList

class NavigationOverlayView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : ViewPager(context, attrs), Navigator.OnManeuverChangedListener {

    companion object {
        /**
         * Helper method to render a maneuverPair.  This is used by the [NavigationOverlayView] and
         * the [RouteSummaryFragment].
         */
        fun renderManeuver(maneuverPair: ManeuverPair,
                           nextDirectionIcon: ImageView,
                           nextDirectionText: TextView,
                           displayHelper: ManeuverDisplayHelper) {
            if (maneuverPair.turnManeuver != null) {
                nextDirectionIcon.setImageResource(displayHelper.getImageResourceForDirection(nextDirectionIcon.context, maneuverPair.turnManeuver!!))

                val textDirection = displayHelper.stringForDirection(nextDirectionText.context, maneuverPair.turnManeuver)
                val textDistance = displayHelper.distanceForDirection(nextDirectionText.context, maneuverPair.mainManeuver, " in")

                val spannableTextDirection = textDirection.toSpannableText(nextDirectionText.context, R.style.DirectionTextAppearance)
                val spannableTextDistance = textDistance.toSpannableText(nextDirectionText.context, R.style.DistanceTextAppearance)

                val spannableBuilder = SpannableStringBuilder().append(spannableTextDirection).append(spannableTextDistance)
                val styledText = spannableBuilder.subSequence(0, spannableBuilder.length)

                nextDirectionText.text = styledText
            } else {
                nextDirectionIcon.setImageResource(displayHelper.getImageResourceForDirection(nextDirectionText.context, maneuverPair.mainManeuver!!))

                val textDirection = displayHelper.stringForDirection(nextDirectionText.context, maneuverPair.mainManeuver)
                val textDistance = displayHelper.distanceForDirection(nextDirectionText.context, maneuverPair.mainManeuver, " for")

                val spannableTextDirection = textDirection.toSpannableText(nextDirectionText.context, R.style.DirectionTextAppearance)
                val spannableTextDistance = textDistance.toSpannableText(nextDirectionText.context, R.style.DistanceTextAppearance)

                val spannableBuilder = SpannableStringBuilder().append(spannableTextDirection).append(spannableTextDistance)
                val styledText = spannableBuilder.subSequence(0, spannableBuilder.length)

                nextDirectionText.text = styledText
            }
        }
    }

    private lateinit var navigator: Navigator
    private lateinit var adapter: ManeuverPagerAdapter

    private val pageChangeListener = object : ViewPager.SimpleOnPageChangeListener() {
        override fun onPageSelected(position: Int) {
            val pair = adapter.getItem(position)
            navigator.setCurrentManeuver(pair.mainPos)
        }
    }

    fun setNavigator(navigator: Navigator) {
        this.navigator = navigator
        navigator.addOnManeuverChangedListener(this)

        val pairs = ArrayList<ManeuverPair>()
        val maneuvers = this.navigator.maneuvers
        var i = 0
        while (i < maneuvers.size) {
            val pair = ManeuverPair()
            pair.mainPos = i
            pair.mainManeuver = maneuvers[i]

            if (i + 1 < maneuvers.size) {
                val next = maneuvers[i + 1]
                if (next.isTurnManeuver || next.isPortalManeuver) {
                    pair.turnPos = i + 1
                    pair.turnManeuver = next
                } else {
                    i--
                }
            }

            pairs.add(pair)
            i += 2
        }

        adapter = ManeuverPagerAdapter(navigator)
        setAdapter(adapter)
        adapter.setManeuvers(pairs)
        addOnPageChangeListener(pageChangeListener)
        currentItem = 0
        navigator.setCurrentManeuver(0)

    }

    override fun onManeuverChanged(navigator: Navigator, position: Int) {
        for (i in 0 until adapter.count) {
            val pair = adapter.getItem(i)
            if (pair.mainPos == position || pair.turnPos == position) {
                currentItem = i
                return
            }
        }
    }

    override fun onRouteSnapFailed() {
        PwLog.e("NavigationOverlayView", "Off route")
    }

    class ManeuverPair {
        internal var mainPos: Int = 0
        internal var turnPos: Int = 0
        internal var mainManeuver: RouteManeuverOptions? = null
        internal var turnManeuver: RouteManeuverOptions? = null
    }

    private class ManeuverPagerAdapter internal constructor(private val navigator: Navigator) : PagerAdapter() {
        private val displayHelper: ManeuverDisplayHelper = ManeuverDisplayHelper()
        private val maneuvers = ArrayList<ManeuverPair>()

        override fun getCount(): Int = maneuvers.size

        internal fun setManeuvers(maneuvers: Collection<ManeuverPair>) {
            this.maneuvers.clear()
            this.maneuvers.addAll(maneuvers)
            notifyDataSetChanged()
        }

        internal fun getItem(position: Int): ManeuverPair = maneuvers[position]

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val context = container.context
            val v = LayoutInflater.from(context).inflate(R.layout.item_maneuver, container, false)
            val nextDirectionIcon = v.findViewById<ImageView>(R.id.next_direction_icon)
            val nextDirectionText = v.findViewById<TextView>(R.id.next_direction_text)

            val currentManeuverPair = maneuvers[position]

            renderManeuver(
                    maneuverPair = currentManeuverPair,
                    nextDirectionText = nextDirectionText,
                    nextDirectionIcon = nextDirectionIcon,
                    displayHelper = displayHelper
            )

            container.addView(v)
            return v
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            container.removeView(`object` as View)
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean = view == `object`
    }

    fun getManeuverPair(): ManeuverPair {
        return adapter.getItem(currentItem)
    }
}

fun String.toSpannableText(context: Context, styleResourceId: Int): SpannableString {
    return SpannableString(this).apply {
        setSpan(TextAppearanceSpan(context, styleResourceId), 0, this.length, 0)
    }
}
